### PR TITLE
Ports paradise's modular buildmode

### PR DESCRIPTION
**Pull request by marlyn-x86**
Whoops, let this sit too long. I've squashed and rebased to fix the conflicts. Repost of #39598
:cl: Crazylemon
admin: "Fill" Buildmode, "Area Edit" Buildmode, "Boom" Buildmode
refactor: Buildmode is now in modules instead of defining new modes by extra branches of switch statements.
imageadd: Switches buildmode button icons to something with a slightly more refined background.
admin: Buildmode now switches between modes using a discrete switcher, instead of by repeatedly clicking.
/:cl:
Ports ParadiseSS13/Paradise#9291

Features it includes:
+ Datumized buildmode sub-modes, to make the system more extensible.
+ A more user-friendly mode and dir-switching system.
+ The following buildmodes:
    + "Fill" Buildmode
    + "Area Edit" Buildmode
    + "Boom" Buildmode

Things I did not port due to wanting to get the simple stuff through first:
+ ["Atmos" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/atmos.dm)
+ ["Map Saving" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/save.dm)
+ ["Door Link" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/link.dm)

Pictures of the new mode switching system:

tgstation_bm_modeswitch

tgstation_bm_dirswitch

### CODE REVIEW

File: code/modules/buildmode/buildmode.dm [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	QDEL_NULL(mode)
+	QDEL_LIST(modeswitch_buttons)
+	QDEL_LIST(dirswitch_buttons)
+	return ..()
+
+/datum/buildmode/proc/post_login()
+	// since these will get wiped upon login
+	holder.screen += buttons
+
+/datum/buildmode/proc/post_logout()
```
Comment (Crazylemon64) [inline code/modules/buildmode/buildmode.dm]: @AnturK I created it as I added callbacks to the `player_details` datum, and did not actually find a use for it. I'll remove it now.

Comment (Crazylemon64) [reply code/modules/buildmode/buildmode.dm]: Consequently, should I remove the entire `player_details` logout (not login) callback system? After this removal, nothing will utilize it.

**Review by AnturK — CHANGES_REQUESTED** [review]
Some minor stuff and this should be good to go.

**Comment by Crazylemon64** [comment]
@AnturK Changes made

**Pull request merged**